### PR TITLE
Cmr 566

### PR DESCRIPTION
**Pull request by dseelig**

### CODE REVIEW

File: src/main/java/de/qucosa/oai/provider/api/dissemination/DisseminationApi.java [hunk 1]
```diff
@@ -3,10 +3,13 @@
 import com.fasterxml.jackson.databind.ObjectMapper;
 import de.qucosa.oai.provider.persitence.Dao;
 import de.qucosa.oai.provider.persitence.dao.postgres.DisseminationDao;
+import de.qucosa.oai.provider.persitence.exceptions.DeleteFailed;
```
Comment (claussni) [inline src/main/java/de/qucosa/oai/provider/api/dissemination/DisseminationApi.java]: Typo: Package name should be `persistence`. Renaming it is an easy task using the refactoring tools.

File: src/main/java/de/qucosa/oai/provider/api/format/FormatApi.java [hunk 2]
```diff
@@ -55,12 +58,12 @@ public T getInputData() {
         return inputData;
     }
 
-    public Format saveFormat(Format format) throws SQLException {
-        return (Format) dao.save(format);
+    public Format saveFormat(Format format) throws SaveFailed {
+        return (Format) dao.saveAndSetIdentifier(format);
```
Comment (claussni) [inline src/main/java/de/qucosa/oai/provider/api/format/FormatApi.java]: Something to consider: These one-line save methods just add another level of indirection and additional (dangerous) class casting. Maybe we can get rid of them somewhen.

File: src/main/java/de/qucosa/oai/provider/config/ApplicationConfig.java [hunk 3]
```diff
@@ -51,24 +51,8 @@ public RestTemplate restTemplate() {
...
-        return api;
-    }
-
-    @Bean
-    public Dao setDao() throws SQLException, PropertyVetoException {
-        Dao setDao = new SetDao<Set>();
-        setDao.setConnection(dataSource());
-        return setDao;
+    public Dao<Set> setDao() throws PropertyVetoException, SQLException {
+        return (Dao<Set>) new SetDao(dataSource().getConnection());
```
Comment (claussni) [inline src/main/java/de/qucosa/oai/provider/config/ApplicationConfig.java]: Tip: If you can construct a bean for the Connection objects and pass it to all the other beans as a dependency, you don't need the potentially dangerous `dataSource().getConnection()` call.

File: src/main/java/de/qucosa/oai/provider/persitence/Dao.java [hunk 4]
```diff
@@ -1,33 +1,34 @@
...
+    T findById(String id) throws NotFound;
 
-    Treturn findByMultipleValues(String clause, String... values) throws SQLException;
+    Collection<T> findByPropertyAndValue(String property, String value) throws NotFound;
 
-    List<Treturn> findAllByColumnAndValue(String column, Tparam value) throws SQLException;
+    T findByMultipleValues(String clause, String... values) throws NotFound;
 
-    Treturn delete(String column, Tparam ident, boolean value) throws SQLException;
+    int delete(String column, String ident, boolean value) throws DeleteFailed;
```
Comment (claussni) [inline src/main/java/de/qucosa/oai/provider/persitence/Dao.java]: Inconsistent API?
In line 23 on can look up an object using an identifier of type `String`. Shouldn't the `delete()` method also return `String`?

File: src/main/java/de/qucosa/oai/provider/persitence/dao/postgres/DisseminationDao.java [hunk 5]
```diff
@@ -13,135 +16,158 @@
...
+                dissemination.setLastmoddate(resultSet.getTimestamp("lastmoddate"));
+                dissemination.setXmldata(resultSet.getString("xmldata"));
+            }
 
-    @Override
-    public List<Dissemination> findAllByColumnAndValue(String column, Tparam value) throws SQLException {
-        return null;
+            return (T) dissemination;
+        } catch (SQLException e) {
+            throw new NotFound("Connat found dissemination.", e);
```
Comment (claussni) [inline src/main/java/de/qucosa/oai/provider/persitence/dao/postgres/DisseminationDao.java]: Typo: It should be "Cannot find dissemination" or "No dissemination found".

File: src/main/java/de/qucosa/oai/provider/persitence/dao/postgres/RecordDao.java [hunk 6]
```diff
@@ -11,139 +14,164 @@
...
+        int deletedRows = 0;
+
+        try {
+            PreparedStatement ps = connection.prepareStatement(sql);
+            ps.setBoolean(1, value);
+            ps.setString(2, ident);
+            deletedRows = ps.executeUpdate();
+
+            if (deletedRows == 0) {
+                throw new DeleteFailed("Record mark as deleted failed, no rwos affected.");
```
Comment (claussni) [inline src/main/java/de/qucosa/oai/provider/persitence/dao/postgres/RecordDao.java]: Typo: Should just be "Failed to mark record as deleted."

File: src/main/java/de/qucosa/oai/provider/persitence/dao/postgres/SetDao.java [hunk 7]
```diff
@@ -17,208 +20,245 @@
...
+                    set.setIdentifier(resultSet.getLong(1));
+                    set.setSetSpec(resultSet.getString("setspec"));
+                    set.setSetName(resultSet.getString("setname"));
+                    set.setSetDescription(resultSet.getString("setdescription"));
+                    set.setDeleted(resultSet.getBoolean("deleted"));
+                    sets.add(set);
+                } while(resultSet.next());
+            }
+        } catch (SQLException e) {
+            e.printStackTrace();
```
Comment (claussni) [inline src/main/java/de/qucosa/oai/provider/persitence/dao/postgres/SetDao.java]: Don't print stack traces.

File: src/main/java/de/qucosa/oai/provider/persitence/dao/postgres/SetDao.java [hunk 8]
```diff
@@ -17,208 +20,245 @@
...
+            connection.commit();
+
+            if (deletedRows == 0) {
+                throw new DeleteFailed("Set mark as deleted failed, no rwos affected.");
+            }
 
-        ps.close();
+            ps.close();
+        } catch (SQLException e) {
+            e.printStackTrace();
```
Comment (claussni) [inline src/main/java/de/qucosa/oai/provider/persitence/dao/postgres/SetDao.java]: !!

File: src/test/java/de/qucosa/oai/provider/dao/RecordTestDao.java [hunk 9]
```diff
@@ -90,37 +92,33 @@ public Record findByColumnAndValue(String column, Tparam value) throws SQLExcept
...
-    }
+            return (Collection<T>) records;
+        } catch (IOException ignore) {
 
-    @Override
-    public Record findByMultipleValues(String clause, String... values) throws SQLException {
-        return null;
+        }
+
+        throw new NotFound("Cannot found record data row.");
```
Comment (claussni) [inline src/test/java/de/qucosa/oai/provider/dao/RecordTestDao.java]: Typo: Should be "Cannot find record data."

File: src/test/java/de/qucosa/oai/provider/dao/RecordTestDao.java [hunk 10]
```diff
@@ -130,25 +128,23 @@ public Record delete(String column, Tparam ident, boolean value) throws SQLExcep
                 i++;
 
                 if (!node.has(column)) {
-                    throw new SQLException(column + " not found in records table.");
+                    throw new DeleteFailed(column + " not found in records table.");
```
Comment (claussni) [inline src/test/java/de/qucosa/oai/provider/dao/RecordTestDao.java]: ..."find", not "found"

File: src/test/java/de/qucosa/oai/provider/dao/RecordTestDao.java [hunk 11]
```diff
@@ -130,25 +128,23 @@ public Record delete(String column, Tparam ident, boolean value) throws SQLExcep
...
                 }
 
                 if (node.get(column).asText().equals(ident)) {
-                    record = om.readValue(node.toString(), Record.class);
-                    record.setRecordId(Long.valueOf(i));
-                    record.setDeleted(value);
-                    return record;
+//                    Record record = om.readValue(node.toString(), Record.class);
+                    return 1;
```
Comment (claussni) [inline src/test/java/de/qucosa/oai/provider/dao/RecordTestDao.java]: Magic number 1?

File: src/test/java/integration/DisseminationIT.java [hunk 12]
```diff
@@ -81,23 +91,55 @@ public void Find_dissemination_by_multiple_criterias() throws SQLException {
...
-        Dissemination dissemination = disseminationApi.findByMultipleValues(
-                "id_format=%s AND id_record=%s",
-                String.valueOf(format.getFormatId()), "oai:example:org:qucosa:55887");
+    public void Mark_dissemination_as_deleted() throws DeleteFailed {
+        Format format = null;
+
+        try {
+            format = (Format) formatApi.find("mdprefix", "oai_dc").iterator().next();
+        } catch (NotFound notFound) {
+            notFound.printStackTrace();
```
Comment (claussni) [inline src/test/java/integration/DisseminationIT.java]: You don't need to print stack traces ever. Not even in tests.

File: src/test/java/de/qucosa/oai/provider/dao/RecordTestDao.java (hunk 11 above)
Comment (dseelig) [reply src/test/java/de/qucosa/oai/provider/dao/RecordTestDao.java]: If find column and match ident value (this parameters will rename in a other branch) than returns 1 for successful delete in test.

File: src/main/java/de/qucosa/oai/provider/persitence/Dao.java (hunk 4 above)
Comment (dseelig) [reply src/main/java/de/qucosa/oai/provider/persitence/Dao.java]: Edit this in a other branch.

File: src/main/java/de/qucosa/oai/provider/persitence/dao/postgres/DisseminationDao.java (hunk 5 above)
Comment (dseelig) [reply src/main/java/de/qucosa/oai/provider/persitence/dao/postgres/DisseminationDao.java]: Changed.

File: src/main/java/de/qucosa/oai/provider/persitence/dao/postgres/SetDao.java (hunk 8 above)
Comment (dseelig) [reply src/main/java/de/qucosa/oai/provider/persitence/dao/postgres/SetDao.java]: Remove stack trace print and add throws delete failed exception.

File: src/test/java/de/qucosa/oai/provider/dao/RecordTestDao.java (hunk 9 above)
Comment (dseelig) [reply src/test/java/de/qucosa/oai/provider/dao/RecordTestDao.java]: Changed.

File: src/test/java/de/qucosa/oai/provider/dao/RecordTestDao.java (hunk 10 above)
Comment (dseelig) [reply src/test/java/de/qucosa/oai/provider/dao/RecordTestDao.java]: Changed.

File: src/test/java/integration/DisseminationIT.java (hunk 12 above)
Comment (dseelig) [reply src/test/java/integration/DisseminationIT.java]: Remove stack trace print and add throws not found exception.

File: src/main/java/de/qucosa/oai/provider/api/dissemination/DisseminationApi.java (hunk 1 above)
Comment (dseelig) [reply src/main/java/de/qucosa/oai/provider/api/dissemination/DisseminationApi.java]: The package is renamed.

File: src/main/java/de/qucosa/oai/provider/api/format/FormatApi.java (hunk 2 above)
Comment (dseelig) [reply src/main/java/de/qucosa/oai/provider/api/format/FormatApi.java]: This problem resolved in a another branch.

File: src/main/java/de/qucosa/oai/provider/config/ApplicationConfig.java (hunk 3 above)
Comment (dseelig) [reply src/main/java/de/qucosa/oai/provider/config/ApplicationConfig.java]: This problem becomes resolved in a another branch.

**Pull request merged**